### ISSUE TITLE
Documentation missing

### ISSUE THREAD

**Issue by nadavsinai** [randing89/ngrequire-webpack-plugin#1]
Hi, I've been trying to use your plugin for a big project I am supporting, trying to covert it to modern FED using webpack and ES6.

I've been trying to understand your code from the test and source but had not luck in making it work.
I think the documentation ought to be improved and I am willing to help (by a PR) if I could get it working for me.
The entry file for webpack declares the main module.it depends on about 50 sub modules which  are all found in the modules folder, and since the tests reside in each module the gloub pattern i wrote was this:

['public/modules/_/_.js',
            'public/modules/_/_[!tests]_/_.js',
            'public/modules/_/_[!tests]_/__/_.js',
            'babel!public/modules/_/_[!tests]_/_.es6']

notice that I am using babel loader for es6 (found out the hard way how to use it...);

webpack runs fine, but does not emit my application files into the output file,
any idea?

**Comment by randing89** [comment]
Can you paste your webpack config?
Thanks

Regards,
Ran Ding

**Comment by nadavsinai** [comment]
sure:

``` javascript
var path = require("path");
var webpack = require("webpack");
var config = require('./config/config');
var ngRequirePlugin = require('ngrequire-webpack-plugin');
module.exports = {
    cache: true,
    context: path.resolve('public'),
    entry: {
        application: ['./application.js']
    },
    output: {
        path: path.join(global.basePath, 'public', "assets", 'js'),
        publicPath: "assets/js/",
        filename: "[name].js",
        chunkFilename: "[chunkhash].js"
    },
    module: {
        loaders: [
            // required to write "require('./style.css')"
            {test: /\.css$/, loader: "style-loader!css-loader"},

            // required for bootstrap icons
            {test: /\.woff$/, loader: "url-loader?prefix=font/&limit=5000&mimetype=application/font-woff"},
            {test: /\.ttf$/, loader: "file-loader?prefix=font/"},
            {test: /\.eot$/, loader: "file-loader?prefix=font/"},
            {test: /\.svg$/, loader: "file-loader?prefix=font/"},
            {test: /\.es6$/, loader: "babel", exclude: /(node_modules)/},
        ]
    },
    devtool: 'source-map',
    resolve: {
        modulesDirectories: ["public/lib", "node_modules"],
        extensions: ["", '.js', '.es6'],
        alias: {}
    },
    plugins: [
        new webpack.ProvidePlugin({
            // Automtically detect jQuery and $ as free var in modules
            // and inject the jquery library
            // This is required by many jquery plugins
            jQuery: "jquery",
            //angular: "angular",
            moment: "moment",
            ZeroClipboard: "zeroclipboard",
            _: "lodash",
            Highcharts: "highcharts-commonjs",
            tinycolor: 'tinycolor',
            MobileDetect: 'mobile-detect',
            scrollMonitor: 'scrollMonitor',
            matchMedia: 'matchmedia',
            $: "jquery",
            introJs: 'intro.js'
        }),
        new webpack.DefinePlugin({
            ENVIRONMENT: process.env.NODE_ENV
        }),
        new ngRequirePlugin([
            'public/modules/application.js',
            'public/modules/*/*.js',
            'public/modules/*/*[!tests]*/*.js',
            'public/modules/*/*[!tests]*/**/*.js',
            'babel!public/modules/*/*[!tests]*/*.es6'
            //    //'public/modules/*/*[!tests]*/**/*(.js|.es6)'
        ]),
        new webpack.optimize.DedupePlugin()
    ]
};
```

**Comment by randing89** [comment]
I guess you need to write something like this.

```
new ngRequirePlugin([
    'public/modules/application.js',
    'public/modules/**/*.js',
    '!public/modules/*/tests/**/*.js'
])
```

The module path only support 3 operations:

```
* = Match any file
** = Match recursively
! = Exclude the file that matching the pattern
```

**Comment by nadavsinai** [comment]
Hey, thanks for the help,

I got this working now, but the plugin is erorring on unknown providers, I
got plenty of 3rd party providers in lib folder which I didn't let it
crawl. (like Restangular, angular Translate etc..)
If I do let it cawrl libs - then it fails on the error of having more than
one module per file (which happens in a lot of angularware)
is there a way to exclude libs?

Thanks

Nadav

**Comment by randing89** [comment]
Hi

In the project I am currently working on we have convention to only define a controller/directive/factory in one file.
This plugin was used to handle ~300 angular files. We found it makes the project more organised. 

We don’t index third party components. For 2 reasons:
1. Third party libraries just have too many ways to define a provider. It is hard to handle all of them.
2. Angular doesn’t handle namespace in module correctly, 2 providers with same name will get injected unpredictably.

When using the third party ones, I will just include them in the traditional way (require(xxx) + specify the namespace + DI as normal).
For other providers we only do DI and the plugin handles everything else for use.

BTW: The plugin will handle the module creation and retrieval automatically, using “angular.module('myModule', [])” on multiple places will not create error.

You are welcome to pull request to make it support multiple providers in a same file.
But currently I have no plan to support that :)

Thanks.

Ran Ding
randing89@gmail.com

**Comment by nadavsinai** [comment]
Hi and sorry for the delay, I was occupied with many other projects, I appreciate you quick reply and your  time.
I was trying to use your plugin to convert a traditional global variable angular,  script loading  -> concated production based app to webpack without going through the 600~ js files and requiring the relevant files (including libs).
I know see this will not be so easy.
I did manage to have some progress diving into your code , adding the options argument like this :
    new ngRequirePlugin({
            include: [
                'public/application.js',
                'public/modules/_/__/_.js',
                'babel!public/modules/_/__/_.es6',
                '!public/modules/_/tests/__.js',
                '!public/modules/_/tests/*_.es6'
            ], options: {ignoreProviders: [/._/]}
        })
basically allowing it to run over my application and ignore the unknown providers from libs.
So as my initial issue suggested, what I mainly wanted to improve ( and understand) is the documentation of your plugin.
Unfortunately I must leave this side project as it is now for lack of time, perhaps it could still happen in the future that I would fork your plugin and continue the development, issuing a PR when I finish, but for now I must attend other projects. You can close this issue for now, however this issue itself now provides some documentation for those that want to work with the options argument.
